### PR TITLE
[flang][openacc] Support allocatable and pointer array in private recipe

### PR DESCRIPTION
**Pull request by clementval**
Add support for pointer and allocatable arrays in private clause.

### CODE REVIEW

File: flang/lib/Lower/OpenACC.cpp [hunk 1]
```diff
@@ -381,7 +394,8 @@ static void genPrivateLikeInitRegion(mlir::OpBuilder &builder, RecipeOp recipe,
       }
     }
   } else if (auto boxTy = mlir::dyn_cast_or_null<fir::BaseBoxType>(ty)) {
-    if (!mlir::isa<fir::SequenceType>(boxTy.getEleTy()))
+    mlir::Type innerTy = extractSequenceType(boxTy);
+    if (!mlir::isa<fir::SequenceType>(innerTy))
```
Comment (vzakhari) [inline flang/lib/Lower/OpenACC.cpp:398]: nit: should `(!innerTy)` be enough?

File: mlir/lib/Dialect/OpenACC/IR/OpenACC.cpp [hunk 2]
```diff
@@ -435,7 +435,7 @@ static LogicalResult verifyInitLikeSingleArgRegion(
 LogicalResult acc::PrivateRecipeOp::verifyRegions() {
   if (failed(verifyInitLikeSingleArgRegion(*this, getInitRegion(),
                                            "privatization", "init", getType(),
-                                           /*verifyYield=*/true)))
+                                           /*verifyYield=*/false)))
```
Comment (vzakhari) [inline mlir/lib/Dialect/OpenACC/IR/OpenACC.cpp:438]: nit: maybe we can use `hlfir::getFortranElementOrSequenceType` in the verifier to check that the sequence/scalar type of the recipe and the yield match vs just dropping verification completely.

**Review by vzakhari — APPROVED** [review]
LGTM, just a couple of minor remarks.

File: flang/lib/Lower/OpenACC.cpp (hunk 1 above)
Comment (clementval) [reply flang/lib/Lower/OpenACC.cpp:398]: Yes. I'll update it.

File: mlir/lib/Dialect/OpenACC/IR/OpenACC.cpp (hunk 2 above)
Comment (clementval) [reply mlir/lib/Dialect/OpenACC/IR/OpenACC.cpp:438]: Unfortunately, the verifier is in the core MLIR and there is no notion of FIR type here. Maybe the element type could be retrieve through an interface somehow.

**Review by razvanlupusoru — APPROVED** [review]
LGTM

File: mlir/lib/Dialect/OpenACC/IR/OpenACC.cpp (hunk 2 above)
Comment (vzakhari) [reply mlir/lib/Dialect/OpenACC/IR/OpenACC.cpp:438]: Right, it looks like a type interface that can be attached to FIR types might be useful here.  I am okay with having verification disabled for the time being.

**Pull request merged**